### PR TITLE
feature(cli): add vscode compatibility handler to inspector proxy

### PR DESCRIPTION
**Pull request by byCedric**
# Why

Builds upon #21449

This adds a handler that makes sure the `vscode-js-debug` extension can be used with the custom inspector proxy. These changes should NOT interfere with the other debuggers.

# How

- Added vscode compatibility handler
  - Add missing `description` to function nodes, as required by vscode-js-debug
    _(before https://github.com/microsoft/vscode-js-debug/issues/1583)_
  - Respond to `Debugger.getPossibleBreakpoints`, Hermes doesn't respond to this message
- ~~Filter out "useless" `callFrames` when debugging~~ 
  **Moved back to PR, we need to limit filtering**
  → https://github.com/expo/expo/pull/21663
- Avoid Hermes binding breakpoints to wrong locations, before sourcemap is loaded 
  → https://github.com/expo/expo/pull/21631
- Avoid Hermes crashing by disabling `Runtime.getProperties` on `symbol` types
  → https://github.com/expo/expo/pull/21639

# Test Plan

- See tests
- ~~Test project with vscode, see [this repository](https://github.com/byCedric/vscode-js-debug-issue-1583) for a test project.~~
- See https://github.com/expo/expo/pull/21641 to test `apps/bare-expo`

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { RuntimeGetProperties, VscodeCompatHandler } from '../VscodeCompat';
+
+describe(VscodeCompatHandler, () => {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts:3]: top-level describe isn't needed and adds unneeded complexity to mock resetting.

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/VscodeCompat.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import type { Protocol } from 'devtools-protocol';
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/VscodeCompat.ts:1]: Assuming this is in devDependencies already?

**Review by EvanBacon — APPROVED** [review]

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/VscodeCompat.ts (hunk 2 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/VscodeCompat.ts:1]: Yep! See: https://github.com/expo/expo/pull/21449/files#diff-00b4bb0f6ea42a3d74201ea46359ab71c5085d6a17f80894dfc24d8614420018R135

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts (hunk 1 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts:3]: Sure, I can remove it.

Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts:3]: Let me remove it after I merged a few follow-up PRs like: #21566, else I'll be rebasing into oblivion

Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts]: done

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/handlers/VscodeCompat.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    // Hermes adds traces of JSI to the callFrames, which are refering to native code.
+    // It doesn't make sense to show these to the user, so we filter them out.
+    if ('method' in message && message.method === 'Debugger.paused') {
+      message.params.callFrames = message.params.callFrames.filter(
+        (frame) =>
+          frame.location.scriptId !== HERMES_INVALID_SCRIPT_ID &&
+          frame.functionName !== HERMES_NATIVE_FUNCTION_NAME
+      );
```
Comment (Kudo) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/handlers/VscodeCompat.ts:92]: would be good if we could keep at least one frame here for developers to know the crashes is somehow from native side.

Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/handlers/VscodeCompat.ts:92]: Do you have an example or some code snippet that triggers the crash? If so, I can limit this filtering to breakpoint `callFrames` only, leaving the crash `callFrames` as is. Does that solve the concern?

Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/handlers/VscodeCompat.ts:89]: It's not a required functionality to make it work, so for now I'll just disable this code to not-block the exception stacks.

**Comment by byCedric** [comment]
@Kudo @EvanBacon I'm merging this to main, with the exception of the `callFrame` filtering. See: https://github.com/expo/expo/pull/21560#discussion_r1131270155

**Pull request merged**